supabase/functions/_script/backup_and_cleanup.ts

```diff
@@ -0,0 +1,89 @@
+import { createClient } from 'https://esm.sh/@supabase/supabase-js@^2.2.3';
+import 'https://deno.land/x/dotenv/load.ts';
+import { r2 } from '../_utils/r2';
+import type { Database } from '../_utils/supabase.types.ts';
+import * as fs from 'https://deno.land/std/fs/mod.ts';
+
+const supabaseUrl = Deno.env.get('SUPABASE_URL') || '***';
+const supabaseServiceRoleKey = Deno.env.get('SUPABASE_SERVICE_ROLE_KEY') || '***';
+const backupFolder = 'backup';
+
+function useSupabase() {
+  const options = {
+    auth: {
+      autoRefreshToken: true,
+      persistSession: true,
+      detectSessionInUrl: false,
+    },
+  };
+  return createClient<Database>(supabaseUrl, supabaseServiceRoleKey, options);
+}
+
+async function backupAndDeleteOldEntries() {
+  try {
+    if (!await fs.exists(backupFolder)) {
+      await fs.ensureDir(backupFolder);
+    }
+
+    const supabase = useSupabase();
+    const daysToKeep = 30;
+    const pageSize = 1000;
+
+    const currentDate = new Date();
+    const cutoffDate = new Date(currentDate.getTime() - daysToKeep * 24 * 60 * 60 * 1000);
+
+    let offset = 0;
+    let hasMoreData = true;
+
+    while (hasMoreData) {
+      const { data: oldEntries, error: queryError } = await supabase
+        .from<Database['public']['Tables']['app_versions']['Row']>('app_versions')
+        .select()
+        .lt('updated_at', cutoffDate.toISOString())
+        .range(offset, offset + pageSize - 1);
+
+      if (queryError) {
+        console.error('Error querying the Supabase table:', queryError.message);
+        break;
+      }
+
+      if (!oldEntries || oldEntries.length === 0) {
+        hasMoreData = false;
+        break;
+      }
+
+      const backupData = JSON.stringify(oldEntries);
+      const backupFilename = `backup-${currentDate.toISOString()}-page-${offset / pageSize}.json`;
+      const backupPath = `${backupFolder}/${backupFilename}`;
+
+      try {
+        await r2.upload(backupPath, new TextEncoder().encode(backupData));
+        console.log(`Backup saved to R2: ${backupPath}`);
+      } catch (backupError) {
+        console.error('Error saving backup to R2:', backupError);
+        throw new Error('Backup to R2 failed. Stopping execution to prevent data loss.');
+      }
+
+      const { error: deleteError } = await supabase
+        .from('app_versions')
+        .delete()
+        .lt('updated_at', cutoffDate.toISOString());
+
+      if (deleteError) {
+        console.error('Error deleting entries from the Supabase table:', deleteError.message);
+        throw new Error('Deletion from the Supabase table failed. Stopping execution.');
+      } else {
+        console.log('Deletion completed successfully.');
+      }
+
+      offset += pageSize;
+    }
+  } catch (e) {
+    console.error('An error occurred:', e.message);
+  }
+}
+
+backupAndDeleteOldEntries().catch((error) => {
+  console.error('An error occurred:', error.message);
+  Deno.exit(1);
+});
```

Comment:
This catch never gets caught as you're already using the catching in the `backupAndDeleteOldEntries` function. Remove the try catch from  `backupAndDeleteOldEntries`
```suggestion
}

backupAndDeleteOldEntries().then(() => 'Backup logs executed successfully').catch((error) => {
  console.error('An error occurred:', error.message);
  Deno.exit(1);
});
```

Reply:
Even this won't execute as everything in the function itself is wrapped under try & catch.

Can just replace with backupAndDeleteOldEntries().